text/0000-dns-label-process-types.md

```diff
@@ -0,0 +1,139 @@
+
+# Motivation
+
+The images resulting from CNB builds are often deployed to internet connected environments, where it's useful to connect or manage the image lifecycle with an internet routable domain name.
+
+For example, after building a web service image with Cloud Native Buildpacks, developers might want to expose various processes to the internet like static-assets.myapp.com and api.myapp.com. However, process types may currently include `_`, making process types like `Static_Assets` allowed, but not internet routable -- `Static_Assets.myapp.com` is not a valid domain name.
+
+Or, developers might want to manage Kubernetes resources dynamically using process types defined in the resulting image:
+
+```
+kubectl autoscale replicaset static-assets --max=6 --min=2 --cpu-percent=50
+```
+
+However, while CNB allows a `Static_Assets` process type, Kubernetes rejects it as a resource name:
+
+```
+object name does not conform to Kubernetes naming requirements: "Static_Assets"
+```
+
+These incompatibilities make using process types in dynamic scenarios challenging.
+
+# What it is
+
+The current Buildpack specification defines these process type restrictions:
+
+> MUST only contain numbers, letters, and the characters ., \_, and -.
+
+While DNS Labels as defined in RFC 1123 are more restrictive. Some notable differences:
+
+- Underscores (`_`) are not allowed
+- Periods (`.`) are not allowed
+- Dashes ('-') are not allowed as the first or last character
+- Uppercase and lowercase letters are not differentiated
+- Must be less than 63 characters
+
+The buildpack specification will be modified to require process types compatible with RFC 1123 for DNS Labels, and require lowercase letters for compatibility with Kubernetes.
+
+# How it Works
+
+The platform specification will be changed to reject process types that do not comply with lowercased RFC 1123 DNS Labels. When a process type fails to meet this syntax, appropriate error messaging may be given by the platform, for example:
+
+```
+The "Static_Assets" process type must comply with lowercased RFC 1123 DNS label syntax. Underscores are not allowed. Uppercase letters are not allowed."
+```
+
+While uppercase letters are allowed in RFC 1123, uppercase and lowercase are not differentiated -- `API` and `api` are the same. DNS providers and Kubernetes enforce (or coerce) uppercase to lowercase to prevent duplicate entries with different casing. The buildpack spec should also enforce lowercase letters for the same reason.
+
+# Migration
+
+This is a breaking API change, and will require a new Buildpack API version.
+
+Buildpacks specifying process types that do not comply with the stricter requirements proposed in this RFC will need to change the way process types are defined. Buildpacks defining process types dynamically should implement logic to coerce process types (by substitution, filtering and/or truncation) to meet the specification and avoid build failures.
```

Comment:
Can we implement this in two phases? First phase, buildpacks violating this will result in a WARNING message. Second phase, buildpacks violating this will break.

That or have a flag that you can use to control the outcome. i.e. `--strict-proc-names=warn` or `--strict-proc-names=error`.

Either of these options would make the migration process easier & give buildpack authors that don't closely follow the spec changes an easy want to tell that something is going on.

Reply:
The flag would be really nice, if you could opt-into that before updating your buildpack APIs. It would make testing and checking for impacted buildpacks easier.

Reply:
Good idea. Added.